### PR TITLE
drivers: i2c: stm32: add bus recovery support

### LINKED ISSUE #54917: drivers: i2c: support for bus recovery on STM32 platforms
**Is your feature request related to a problem? Please describe.**
I've encountered an issue with multiple STM32F4 boards that operate for several minutes to hours without issue suddenly losing control over their i2C bus. The driver attempts to reset the driver with `stm32_i2c_reset()`, but never regains control of the bus. What I'm encountering is very similar to what's been described in [#27628](https://github.com/zephyrproject-rtos/zephyr/issues/27628).

**Describe the solution you'd like**
Some sort of "9 pulse" bus recovery support added to the STM32 i2c drivers. This already exist for two zephyr platforms while [#46472](https://github.com/zephyrproject-rtos/zephyr/pull/46472) has some talk about adding it to the top-level to all i2c drivers can take advantage of it..

**Additional context**
Currently I have a poorly implemented patch implemented in my i2c_ll_stm32_v1.c file that takes control of the i2c lines as GPIO and recovers the bus. This has been proven to work and I've even shown that manually jumpering SCL to ground can clear the bus as well.

### PR DESCRIPTION
**Pull request by maxmclau**
Add I2C bus recovery support to the STM32 v1 and v2 driver.

The STM32 i2c peripheral does not natively support I2C bus recovery so recovery is performed using GPIO bitbanging. This mirrors the bus recovery implementation for NXP MCUX LPI2C driver.

Fixes: #54917

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
> Merge branch 'main' into stm32-bus-recovery

@maxmclau please rebase your patch on top of the main branch; do not merge the main branch into your PR. We cannot merge PRs with merge commits.

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml [hunk 1]
```diff
@@ -19,3 +19,17 @@ properties:
 
   pinctrl-names:
     required: true
+
+  scl-gpios:
```
Comment (erwango) [inline dts/bindings/i2c/st,stm32-i2c-v1.yaml:23]: I wonder if there is a possibility to extend and use what is proposed [here](https://github.com/zephyrproject-rtos/zephyr/pull/54880/files#r1113056596) to retrieve directly gpio information from the pinctrl definition.
Do you need anything else than gpio port and pin ? (gpio flags ?)

File: drivers/i2c/Kconfig.stm32 [hunk 2]
```diff
@@ -43,4 +43,11 @@ config I2C_STM32_COMBINED_INTERRUPT
 	depends on I2C_STM32_INTERRUPT
 	default y if SOC_SERIES_STM32F0X || SOC_SERIES_STM32G0X || SOC_SERIES_STM32L0X
 
+config I2C_STM32_BUS_RECOVERY
+	bool "Bus recovery support"
+	depends on I2C_STM32_V1 && PINCTRL
```
Comment (erwango) [inline drivers/i2c/Kconfig.stm32:48]: Using I2C requires PINCTRL anyway, so dependecy on PINCTRL is useless.
Then, I wonder about dependency on I2C_STM32_V1. Is that really required ? I don't see anything V1 specific in your proposal.

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml (hunk 1 above)
Comment (maxmclau) [reply dts/bindings/i2c/st,stm32-i2c-v1.yaml:23]: I agree, that's a much cleaner solution than requiring the gpio phandles. I'd prefer to wait until that's implemented somewhere generically accessible as is being discussed

**Comment by maxmclau** [comment]
Waiting on resolution of pinctrl discussion in #54880 to remove separate gpio phandle dependency.

File: drivers/i2c/i2c_ll_stm32.c [hunk 3]
```diff
@@ -460,6 +561,8 @@ static const struct i2c_stm32_config i2c_stm32_cfg_##index = {		\
 	STM32_I2C_IRQ_HANDLER_FUNCTION(index)				\
 	.bitrate = DT_INST_PROP(index, clock_frequency),		\
 	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(index),			\
+	I2C_STM32_SCL_INIT(index)				\
+	I2C_STM32_SDA_INIT(index)				\
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:565]:
```suggestion
	I2C_STM32_SCL_INIT(index)					\
	I2C_STM32_SDA_INIT(index)					\
```

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
> Waiting on resolution of pinctrl discussion in https://github.com/zephyrproject-rtos/zephyr/pull/54880 to remove separate gpio phandle dependency.

Given the initial complexity to support instantiation in those macros, I'd suggest to move on with current proposal and we can get back to it later too avoid excessive delays.
This doesn't prevent to get the current version as close as possible to a mergeable state. Aside from the few comments provided, this looks pretty good to me and close to merge.

**Review by FRASTM — COMMENTED** [review]
LGTM

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml (hunk 1 above)
Comment (Bytegrity) [reply dts/bindings/i2c/st,stm32-i2c-v1.yaml:23]: A remark on the macros introduced in my PR. Their purpose is to perform logical operations through DT based macros on phandle nodes within a pinctrl to provide insights on whether a particular function should be enabled.

What I am gathering is you would require to  know what phandle contains either the `scl` or `sda` function without caring that much about their actual pin and port designators?
 
Trying to get an accurate picture to assess how they could hopefully be useful for your case.

File: drivers/i2c/Kconfig.stm32 (hunk 2 above)
Comment (maxmclau) [reply drivers/i2c/Kconfig.stm32]: Neither are required - removed in latest push

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml (hunk 1 above)
Comment (maxmclau) [reply dts/bindings/i2c/st,stm32-i2c-v1.yaml:23]: I see, these use don't line up as well as I originally thought. Ultimately I'm looking to retrieve just enough information from a pinctrl definition to form a gpio spec for manual pin setting. For this I'm only interested in extracting the pin and port designators - flags should not matter.

**Comment by maxmclau** [comment]
Should be ready for another review.

**Review by erwango — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

**Comment by erwango** [comment]
@maxmclau Please fix compliance issue

**Comment by maxmclau** [comment]
Think all compliance issues have been resolved

**Review by teburd — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**